### PR TITLE
Feat/exchange mfa

### PR DESCRIPTION
**Pull request by cbourgois**
### Description of the Change

5034117 — feat(exchange.mfa): add exchange mfa

replace https://github.com/ovh/manager/pull/1851

ref: MANAGER-3014 

### Related PR : 

- [x] https://github.com/ovh-ux/ovh-api-services/pull/273

### CODE REVIEW

File: packages/manager/apps/dedicated/package.json [hunk 1]
```diff
@@ -119,7 +119,7 @@
     "oclazyload": "^1.1.0",
     "office-ui-fabric-core": "^11.0.0",
     "ovh-angular-responsive-tabs": "^4.0.0",
-    "ovh-api-services": "^9.30.0",
+    "ovh-api-services": "^9.32.0",
```
Comment (marie-j) [inline packages/manager/apps/dedicated/package.json]: Could have been nice to `yarn upgrade-interactive` separately on `develop` branch

File: packages/manager/modules/exchange/src/account/account.service.js [hunk 2]
```diff
@@ -89,6 +91,141 @@ export default class ExchangeAccount {
...
+      .v6()
+      .create(
+        {
+          serviceName,
+          userPrincipalName,
+        },
+        {},
+      )
+      .$promise.then((data) => {
+        this.Exchange.refreshViews('Accounts', 'Tasks');
```
Comment (marie-j) [inline packages/manager/modules/exchange/src/account/account.service.js]: I suppose there is no other way around ? The service shouldn't be triggering a view refresh

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      case 'RESET':
+        accountMfaPromise = this.reset();
+        messages = {
+          success: 'exchange_reset_mfa_success',
+          error: 'exchange_reset_mfa_error',
+        };
+        break;
+      default:
+        this.$scope.resetAction();
+    }
```
Comment (marie-j) [inline packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js]: Maybe we could use something like 
```suggestion
    accountMfaPromise = this[this.action.toLowerCase]() 
    messages = {
    success: `exchange_{this.action.toLowerCase}_mfa_success`
    ... 
```

**Comment by ovh-cds** [comment]
CDS Report test-manager_2#8607.0 ■   
* Stage 1
  * e2e ■   
  * lint ■

**Comment by ovh-cds** [comment]
CDS Report build-monorepository#8607.0 ■   
* 
  * Build ■

File: packages/manager/modules/exchange/src/account/mfa/bulk-create.html [hunk 4]
```diff
@@ -0,0 +1 @@
+<exchange-account-mfa-create></exchange-account-mfa-create>
```
Comment (marie-j) [inline packages/manager/modules/exchange/src/account/mfa/bulk-create.html]: Isn't it possible to just pass the component to a `ui-view` or something like that ?

**Comment by ovh-cds** [comment]
CDS Report test-manager_2#8607.0 ■   
* Stage 1
  * e2e ■   
  * lint ■

File: packages/manager/modules/exchange/src/exchange.service.js [hunk 5]
```diff
@@ -1755,7 +1755,21 @@ export default class Exchange {
           exchangeService: name,
         },
       },
-    ).then(this.getSuccessDataOrReject);
+    );
+  }
+
+  updateExchangeServer(organization, name, data) {
+    return this.services.OvhHttp.put(
```
Comment (marie-j) [inline packages/manager/modules/exchange/src/exchange.service.js]: Cn't it be replaced by a `ovh-api-services` call ?

Comment (cbourgois) [reply packages/manager/modules/exchange/src/exchange.service.js]: Oh I forgot it after `git subtree pull`

File: packages/manager/modules/exchange/src/account/mfa/bulk-create.html (hunk 4 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/account/mfa/bulk-create.html]: Invoked with some `setAction` code. I haven't refactor the routing.

File: packages/manager/modules/exchange/src/account/account.service.js (hunk 2 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/account/account.service.js]: Not yet, `exchange` isn't "full-`uirouter`" yet, and it's currently done like this.

File: packages/manager/apps/dedicated/package.json (hunk 1 above)
Comment (cbourgois) [reply packages/manager/apps/dedicated/package.json]: IMO, only `exchange` require a bump of `ovh-api-services` (and therefore `web`and `dedicated` manager). We don't have to bump `ovh-api-services` right now in other packages as they don't need the features introduced in `ovh-api-services@9.32.0`.

File: packages/manager/modules/exchange/src/exchange.service.js (hunk 5 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/exchange.service.js]: Need to release a new version (https://github.com/ovh-ux/ovh-api-services/pull/273)

File: packages/manager/modules/exchange/src/account/account.service.js (hunk 2 above)
Comment (FredericEspiau) [reply packages/manager/modules/exchange/src/account/account.service.js]: Damn, who rewrote this section two years ago and didn't change that ???

File: packages/manager/modules/exchange/src/account/mfa/delete/index.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import angular from 'angular';
+import component from './delete.component';
+
+const moduleName = 'ExchangeAccountMfaDelete';
+
+angular
+  .module(moduleName, [])
```
Comment (FredericEspiau) [inline packages/manager/modules/exchange/src/account/mfa/delete/index.js]: Don't you use translations, ui kit or routing ? 🤔

File: packages/manager/modules/exchange/src/exchange.html [hunk 7]
```diff
@@ -96,7 +96,13 @@
         data-ng-if="ctrl.shouldOpenWizard && ctrl.hasNoDomain"
     ></exchange-wizard-hosted-creation>
 
-    <div class="modal fade currentAction" id="currentAction" role="dialog">
+    <div
+        class="modal fade currentAction"
+        id="currentAction"
+        tabindex="-1"
+        role="dialog"
+        aria-hidden="true"
```
Comment (FredericEspiau) [inline packages/manager/modules/exchange/src/exchange.html]: Why ?

File: packages/manager/modules/exchange/src/exchange.html [hunk 8]
```diff
@@ -96,7 +96,13 @@
         data-ng-if="ctrl.shouldOpenWizard && ctrl.hasNoDomain"
     ></exchange-wizard-hosted-creation>
 
-    <div class="modal fade currentAction" id="currentAction" role="dialog">
+    <div
+        class="modal fade currentAction"
+        id="currentAction"
+        tabindex="-1"
```
Comment (FredericEspiau) [inline packages/manager/modules/exchange/src/exchange.html]: Why ?

File: packages/manager/modules/exchange/src/exchange.template.js [hunk 9]
```diff
@@ -395,6 +394,18 @@ export default /* @ngInject */ ($templateCache) => {
     'exchange/account/outlook/activate/account-outlook-activate.html',
     accountOutlookActivateTemplate,
   );
+  $templateCache.put(
```
Comment (FredericEspiau) [inline packages/manager/modules/exchange/src/exchange.template.js]: 😅

File: packages/manager/modules/exchange/src/security/security.controller.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
+import clone from 'lodash/clone';
+import get from 'lodash/get';
+import set from 'lodash/set';
+import isNumber from 'lodash/isNumber';
```
Comment (FredericEspiau) [inline packages/manager/modules/exchange/src/security/security.controller.js]: redro lacitebahpla

File: packages/manager/modules/exchange/src/exchange.tabs.controller.js [hunk 11]
```diff
@@ -90,25 +93,28 @@ export default class ExchangeTabsCtrl {
...
-        this.selectedTab = $routerParams.tab.toUpperCase();
+        this.currentTab = this.$routerParams.tab.toUpperCase();
       }
 
-      this.updateScope();
-      this.changeTab(this.selectedTab);
+      this.changeTab(this.currentTab);
+    });
+
+    this.services.$scope.$on('$locationChangeSuccess', () => {
```
Comment (FredericEspiau) [inline packages/manager/modules/exchange/src/exchange.tabs.controller.js]: Why did you change all this part ?

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js (hunk 3 above)
Comment (marie-j) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js]: I was thinking of adapting another OOP pattern which is reflection 😄 
That way we don't have to add another `case` to add another behaviour (though we have to add a function)

File: packages/manager/modules/exchange/src/exchange.tabs.controller.js (hunk 11 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/exchange.tabs.controller.js]: To allow to update tabs when url is updated.

File: packages/manager/apps/dedicated/package.json (hunk 1 above)
Comment (cbourgois) [reply packages/manager/apps/dedicated/package.json]: Updated for all applications using `yarn upgrade-interactive  --latest`

File: packages/manager/modules/exchange/src/exchange.template.js (hunk 9 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/exchange.template.js]: Yes, need a large refactoring for this module.

File: packages/manager/modules/exchange/src/security/security.controller.js (hunk 10 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/security/security.controller.js]: !dexif s'ti

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js (hunk 3 above)
Comment (FredericEspiau) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js]: I know @marie-j, but I think it's possible it could do more harm than good

This articles sums up my opinion about this: https://overreacted.io/goodbye-clean-code/

Code duplication sometimes increases legibility and is easier to maintain than having the wrong abstraction

File: packages/manager/modules/exchange/src/translations/Messages_fr_FR.json [hunk 12]
```diff
@@ -1565,5 +1582,23 @@
   "exchange_wizardHostedCreation_configureDNSZone_manual_dialog_warning": "Attention",
   "exchange_wizardHostedCreation_lastDialog_closeButton": "Finaliser",
   "exchange_table_manage_entry": ": gérer",
-  "sharepoint_activate_which_service_label": "Serveur Exchange à associer"
+  "sharepoint_activate_which_service_label": "Serveur Exchange à associer",
+  "exchange_mfa_description_ENABLE": "Êtes-vous sûr de vouloir activer la double authentification pour le compte <strong>{{account}}</strong>?",
```
Comment (JeremyDec) [inline packages/manager/modules/exchange/src/translations/Messages_fr_FR.json]:
```suggestion
  "exchange_mfa_description_ENABLE": "Êtes-vous sûr de vouloir activer la double authentification pour le compte <strong>{{account}}</strong> ?",
```

File: packages/manager/modules/exchange/src/translations/Messages_fr_FR.json [hunk 13]
```diff
@@ -1565,5 +1582,23 @@
...
+  "exchange_mfa_description_DELETE": "Êtes-vous sûr de vouloir supprimer la double authentification pour le compte <strong>{{account}}</strong>?",
```
Comment (JeremyDec) [inline packages/manager/modules/exchange/src/translations/Messages_fr_FR.json]:
```suggestion
  "exchange_mfa_description_DELETE": "Êtes-vous sûr de vouloir supprimer la double authentification pour le compte <strong>{{account}}</strong> ?",
```

File: packages/manager/modules/exchange/src/security/security.html [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+        <h2
+            class="oui-heading_underline"
+            data-ng-if="ctrl.isMfaAvailable"
+            data-translate="exchange_ACTION_configure_services_label_mfa"
+        ></h2>
+
+        <div class="form-group" data-ng-if="ctrl.isMfaAvailable">
+            <oui-radio-toggle-group model="ctrl.owaMfa" class="float-left">
+                <oui-radio value="true">
+                    {{:: 'spf_value_true' | translate}}
```
[the remark this answers is not in the document]
Comment (cbourgois) [reply packages/manager/modules/exchange/src/security/security.html]: or `span` with `data-translate`...

File: packages/manager/modules/exchange/src/account/mfa/delete/index.js (hunk 6 above)
Comment (cbourgois) [reply packages/manager/modules/exchange/src/account/mfa/delete/index.js]: Of course ;) I'll add them.

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    );
+  }
+
+  submit() {
+    this.isLoading = true;
+
+    let accountMfaPromise = null;
+    let messages = {};
+    switch (this.action) {
+      case 'DELETE':
```
[the remark this answers is not in the document]
Comment (cbourgois) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js]: I don't like to write constant like this : 
```js
export const DELETE = 'DELETE';
```

[the remark this answers is not in the document]
Comment (JeremyDec) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js]: Imo, one interest of constants, is to have a unique point for a stored value to handle it This in order to not having to replace 'DELETE' x times you encounter it in the code, and even worse, to forget one place where you have to replace it. Even if it is to say DELETE = 'DELETE'.

But, it's not a blocking point, so do as you wish

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.html [hunk 16]
```diff
@@ -0,0 +1,32 @@
+<div data-ng-controller="ExchangeAccountMfaCtrl as $ctrl">
```
[the remark this answers is not in the document]
Comment (marie-j) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.html]: @cbourgois No component possible here ?

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (FredericEspiau) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.controller.js]: Imho, I view it better as an `enum`, this makes is a single point of truth about what actions are possible to do

File: packages/manager/modules/exchange/src/account/mfa/account-mfa.html (hunk 16 above)
[the remark this answers is not in the document]
Comment (FredericEspiau) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.html]: *echo*

[the remark this answers is not in the document]
Comment (cbourgois) [reply packages/manager/modules/exchange/src/account/mfa/account-mfa.html]: We could create a component for that but it will not be a _real_ component.

This component will not take parameters (_bindings_), because it needs to use _statefull_ Exchange services to retrieve the current account and the current exchange service.
IMO, it would be necessary to modify the Exchange module in depth for this to bring more sense.

**Comment by marie-j** [comment]
@cbourgois There's conflict

**Comment by cbourgois** [comment]
> @cbourgois There are conflicts

Thanks @frenautvh :)

File: packages/manager/modules/exchange/src/exchange.html (hunk 8 above)
Comment (antleblanc) [reply packages/manager/modules/exchange/src/exchange.html]: It will be fixed. Thanks!
